File: lib/linkage-node/src/gamepad/ps_controller.ts
import { clamp, mapRange } from '../util';
import {
  AxisControl,
  ButtonControl,
  Control,
  EventType
} from './gamepad_manager';

/**
 * Represents the state of a xbox game controller.
 */
export class PsController {
  /**
   * @returns Whether or not the triangle button is pressed.
   */
  triangle = false;

  /**
   * @returns Whether or not the square button is pressed.
   */
  square = false;

  /**
   * @returns Whether or not the cross button is pressed.
   */
  cross = false;

  /**
   * @returns Whether or not the circle button is pressed.
   */
  circle = false;

  /**
   * @returns Whether or not the up button of the D-Pad is pressed.
   */
  dpadUp = false;

  /**
   * @returns Whether or not the down button of the D-Pad is pressed.
   */
  dpadDown = false;

  /**
   * @returns Whether or not the left button of the D-Pad is pressed.
   */
  dpadLeft = false;

  /**
   * @returns Whether or not the right button of the D-Pad is pressed.
   */
  dpadRight = false;

  /**
   * @returns Whether or not the left bumper is pressed.
   */
  leftBumper = false;

  /**
   * @returns Whether or not the right bumper is pressed.
   */
  rightBumper = false;

  /**
   * How far the left trigger has been pressed.
   *
   * @returns A value between 0.0 and 1.0 indicating not pressed at all or fully pressed respectively.
   */
  leftTrigger = 0;

  /**
   * How far the right trigger has been pressed.
   *
   * @returns A value between 0.0 and 1.0 indicating not pressed at all or fully pressed respectively.
   */
  rightTrigger = 0;
  /**
   * The X axis of the left joystick.
   *
   * @returns A value between -1.0 and 1.0 indicating completely to the left and completely to the right respectively. A value of 0.0 means it is horizontally centered.
   */
  leftJoystickX = 0;

  /**
   * The Y axis of the left joystick.
   *
   * @returns A value between -1.0 and 1.0 indicating completely at the bottom and completely at the bottom respectively. A value of 0.0 means it is vertically centered.
   */
  leftJoystickY = 0;

  /**
   * @returns Whether or not the left joystick button is pressed.
   */
  leftJoystickButton = false;

  /**
   * The X axis of the right joystick.
   *
   * @returns A value between -1.0 and 1.0 indicating completely to the left and completely to the right respectively. A value of 0.0 means it is horizontally centered.
   */
  rightJoystickX = 0;

  /**
   * The Y axis of the right joystick.
   *
   * @returns A value between -1.0 and 1.0 indicating completely at the bottom and completely at the bottom respectively. A value of 0.0 means it is vertically centered.
   */
  rightJoystickY = 0;

  /**
   * @returns Whether or not the right joystick button is pressed.
   */
  rightJoystickButton = false;

  /**
   * @returns Whether or not the share button is pressed.
   */
  share = false;

  /**
   * @returns Whether or not the options button is pressed.
   */
  options = false;

  /**
   * @returns Whether or not the home button is pressed.
   */
  home = false;

  public setValue(eventType: EventType, control: Control, value: number): void {
    if (eventType === EventType.AXIS_CHANGED) {
      switch (control as AxisControl) {
        case AxisControl.LEFT_STICK_X:
          this.leftJoystickX = clamp(value / 255, 0, 1);
          break;
        case AxisControl.LEFT_STICK_Y:
          this.leftJoystickY = clamp(value / 255, 0, 1);
          break;
        case AxisControl.RIGHT_STICK_X:
          this.rightJoystickX = clamp(value / 255, 0, 1);
          break;
        case AxisControl.RIGHT_STICK_Y:
          this.rightJoystickY = clamp(value / 255, 0, 1);
          break;
        case AxisControl.DPAD_X:
          this.dpadLeft = false;
          this.dpadRight = false;
          if (value <= -0.5) this.dpadLeft = true;
          else if (value >= 0.5) this.dpadRight = true;
          break;
        case AxisControl.DPAD_Y:
          this.dpadDown = false;
          this.dpadUp = false;
          if (value <= -0.5) this.dpadDown = true;
          else if (value >= 0.5) this.dpadUp = true;
          break;
        default:
          break;
      }
    } else if (eventType === EventType.BUTTON_CHANGED) {
      switch (control as ButtonControl) {
        case ButtonControl.SOUTH:
          this.cross = value >= 127;
          break;
        case ButtonControl.EAST:
          this.circle = value >= 127;
          break;
        case ButtonControl.NORTH:
          this.triangle = value >= 127;
          break;
        case ButtonControl.WEST:
          this.square = value >= 127;
          break;
        case ButtonControl.LEFT_TRIGGER:
          this.leftBumper = value >= 127;
          break;
        case ButtonControl.LEFT_TRIGGER_2:
          this.leftTrigger = clamp(value / 255, 0, 1);
          break;
        case ButtonControl.RIGHT_TRIGGER:
          this.rightBumper = value >= 127;
          break;
        case ButtonControl.RIGHT_TRIGGER_2:
          this.rightTrigger = clamp(value / 255, 0, 1);
          break;
        case ButtonControl.SELECT:
          this.share = value >= 127;
          break;
        case ButtonControl.START:
          this.options = value >= 127;
          break;
        case ButtonControl.MODE:
          this.home = value >= 127;
          break;
        case ButtonControl.LEFT_THUMB:
          this.leftJoystickButton = value >= 127;
          break;
        case ButtonControl.RIGHT_THUMB:
          this.rightJoystickButton = value >= 127;
          break;
        case ButtonControl.DPAD_UP:
          this.dpadUp = value >= 127;
          break;
        case ButtonControl.DPAD_DOWN:
          this.dpadDown = value >= 127;
          break;
        case ButtonControl.DPAD_LEFT:
          this.dpadLeft = value >= 127;
          break;
        case ButtonControl.DPAD_RIGHT:
          this.dpadRight = value >= 127;
          break;
        default:
          break;
      }
    }
  }
}

File: lib/linkage-rs/src/gamepads/ps_controller.rs
use super::gamepad::{AxisControl, ButtonControl, Gamepad, GamepadData};
use crate::gamepads::gamepad::GamepadId;

/// Represents a PlayStation controller.
pub struct PsController {
    gamepad_data: GamepadData,
}

impl Gamepad for PsController {
    /// Creates a new [`PsController`] instance with the given [GamepadData][`crate::gamepads::gamepad::GamepadData`].
    ///
    /// # Parameters
    /// - `gamepad_data`: The [GamepadData][`crate::gamepads::gamepad::GamepadData`] to be associated with the PlayStation controller.
    ///
    /// # Returns
    /// A new [`PsController`] instance.
    fn new(gamepad_data: GamepadData) -> Self {
        Self { gamepad_data }
    }
}

impl PsController {
    /// Returns the [GamepadId][`crate::gamepads::gamepad::GamepadId`] of the PlayStation controller.
    ///
    /// # Returns
    /// The [GamepadId][`crate::gamepads::gamepad::GamepadId`] of the PlayStation controller.
    pub fn id(&self) -> GamepadId {
        self.gamepad_data.gamepad_id()
    }

    /// Returns the state of the triangle button.
    ///
    /// # Returns
    /// `true` if the Triangle button is pressed, `false` otherwise.
    pub fn triangle(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::North as u8)
    }

    /// Returns the state of the square button.
    ///
    /// # Returns
    /// `true` if the square button is pressed, `false` otherwise.
    pub fn square(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::West as u8)
    }

    /// Returns the state of the cross button.
    ///
    /// # Returns
    /// `true` if the cross button is pressed, `false` otherwise.
    pub fn cross(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::South as u8)
    }

    /// Returns the state of the circle button.
    ///
    /// # Returns
    /// `true` if the circle button is pressed, `false` otherwise.
    pub fn circle(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::East as u8)
    }

    /// Returns the state of the dpad's up button.
    ///
    /// # Returns
    /// `true` if the dpad's up button is pressed, `false` otherwise.
    pub fn dpad_up(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::DpadUp as u8)
    }

    /// Returns the state of the dpad's down button.
    ///
    /// # Returns
    /// `true` if the dpad's down button is pressed, `false` otherwise.
    pub fn dpad_down(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::DpadDown as u8)
    }

    /// Returns the state of the dpad's left button.
    ///
    /// # Returns
    /// `true` if the dpad's left button is pressed, `false` otherwise.
    pub fn dpad_left(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::DpadLeft as u8)
    }

    /// Returns the state of the dpad's right button.
    ///
    /// # Returns
    /// `true` if the dpad's right button is pressed, `false` otherwise.
    pub fn dpad_right(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::DpadRight as u8)
    }

    /// Returns the state of the left bumper button.
    ///
    /// # Returns
    /// `true` if the left bumper button is pressed, `false` otherwise.
    pub fn left_bumper(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::LeftTrigger as u8)
    }

    /// Returns the state of the right bumper button.
    ///
    /// # Returns
    /// `true` if the right bumper button is pressed, `false` otherwise.
    pub fn right_bumper(&self) -> bool {
        self.control_button_value(
            &self.gamepad_data.buttons,
            ButtonControl::RightTrigger as u8,
        )
    }

    /// Returns the state of the left trigger.
    ///
    /// # Returns
    /// A value between 0.0 and 1.0 representing not pressed or fully pressed respectively.
    pub fn left_trigger(&self) -> f32 {
        self.control_axis_value(
            &self.gamepad_data.buttons,
            ButtonControl::LeftTrigger2 as u8,
            (0f32, 1f32),
        )
    }

    /// Returns the state of the right trigger.
    ///
    /// # Returns
    /// A value between 0.0 and 1.0 representing not pressed or fully pressed respectively.
    pub fn right_trigger(&self) -> f32 {
        self.control_axis_value(
            &self.gamepad_data.buttons,
            ButtonControl::RightTrigger2 as u8,
            (0f32, 1f32),
        )
    }

    /// Returns the x-axis of the left joystick
    ///
    /// # Returns
    /// A value between -1.0 and 1.0 representing completely left or completely right
    /// respectively with 0.0 being the center.
    pub fn left_joystick_x(&self) -> f32 {
        self.control_axis_value(
            &self.gamepad_data.axis,
            AxisControl::LeftStickX as u8,
            (-1f32, 1f32),
        )
    }

    /// Returns the y-axis of the left joystick
    ///
    /// # Returns
    /// A value between -1.0 and 1.0 representing completely up or completely down
    /// respectively with 0.0 being the center.
    pub fn left_joystick_y(&self) -> f32 {
        self.control_axis_value(
            &self.gamepad_data.axis,
            AxisControl::LeftStickY as u8,
            (-1f32, 1f32),
        )
    }

    /// Returns the state of the left joystick button.
    ///
    /// # Returns
    /// `true` if the left joystick button is pressed, `false` otherwise.
    pub fn left_joystick_button(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::LeftThumb as u8)
    }

    /// Returns the y-axis of the right joystick
    ///
    /// # Returns
    /// A value between -1.0 and 1.0 representing completely up or completely down
    /// respectively with 0.0 being the center.
    pub fn right_joystick_x(&self) -> f32 {
        self.control_axis_value(
            &self.gamepad_data.axis,
            AxisControl::RightStickX as u8,
            (-1f32, 1f32),
        )
    }

    /// Returns the y-axis of the right joystick
    ///
    /// # Returns
    /// A value between -1.0 and 1.0 representing completely up or completely down
    /// respectively with 0.0 being the center.
    pub fn right_joystick_y(&self) -> f32 {
        self.control_axis_value(
            &self.gamepad_data.axis,
            AxisControl::RightStickY as u8,
            (-1f32, 1f32),
        )
    }

    /// Returns the state of the right joystick button.
    ///
    /// # Returns
    /// `true` if the right joystick button is pressed, `false` otherwise.
    pub fn right_joystick_button(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::RightThumb as u8)
    }

    /// Returns the state of share button.
    ///
    /// # Returns
    /// `true` if the share button is pressed, `false` otherwise.
    pub fn share(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::Select as u8)
    }

    /// Returns the state of options button.
    ///
    /// # Returns
    /// `true` if the options button is pressed, `false` otherwise.
    pub fn options(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::Start as u8)
    }

    /// Returns the state of home button.
    ///
    /// # Returns
    /// `true` if the home button is pressed, `false` otherwise.
    pub fn home(&self) -> bool {
        self.control_button_value(&self.gamepad_data.buttons, ButtonControl::Mode as u8)
    }
}
